File: wrappers/android/src/main/kotlin/com/printerdriver/PrinterDriver.kt
package com.printerdriver

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.content.Context
import com.printerdriver.internal.NativeBridge
import com.printerdriver.internal.NativeLogCallback
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * The service (docs/api.md §2): owns connections, queues, the persistent job store.
 * One [PrinterDriver] wraps one pd_driver -- construct with [create], release with
 * [close] (or `use { ... }`, since this implements [AutoCloseable]).
 *
 * Construction can fail (pd_create returns NULL only when its storage directory could
 * not be created), which a Kotlin constructor cannot express -- hence the
 * `create`-then-throw factory instead of a public constructor.
 */
class PrinterDriver private constructor(internal val handle: Long) : AutoCloseable {

    private val closedFlag = AtomicBoolean(false)

    // Backs Printer.send's closure-ergonomics sugar (docs/api.md §12) -- Dispatchers.Main
    // so onProgress/onResult land on the main thread by default, the conventional
    // choice for callback-style Android APIs; see Printer.send's KDoc. A SupervisorJob
    // so one submission's failure/cancellation cannot cancel unrelated ones sharing
    // this scope.
    private val closureScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    // Keyed by printer handle so Printer.bluetoothTransportKind can answer §25's "which
    // of the five paths" question, and so close() can stop every reader thread before
    // pd_destroy frees the handles they feed.
    private val bluetoothTransports = ConcurrentHashMap<Long, BluetoothSppTransport>()

    /** Why the last call on this driver returned NULL/failed. Never null; "" when the
     *  last call succeeded (pd_last_error's own contract, unchanged here). Only
     *  meaningful immediately after a failure -- pd.h: "valid until the next pd_* call
     *  on that same driver." */
    val lastError: String
        get() {
            checkOpen()
            return NativeBridge.driverLastError(handle)
        }

    /** The capability-profile ids [TcpPrinterConfig.profileId] accepts (pd_profile_ids) --
     *  enumerated from the native layer rather than hardcoded here, so this list can
     *  never drift out of sync with what the linked native library actually supports. */
    fun profileIds(): List<String> {
        checkOpen()
        return NativeBridge.profileIds().toList()
    }

    /** Fleet-style: a stable printerId (docs/api.md §2). Throws [PrinterDriverException]
     *  on failure (empty host, unknown [TcpPrinterConfig.profileId]) -- both are
     *  configuration mistakes, not routine runtime conditions. */
    fun addPrinterTcp(config: TcpPrinterConfig): Printer {
        checkOpen()
        val printerHandle = NativeBridge.addPrinterTcp(
            handle, config.printerId, config.host, config.port, config.widthDots,
            config.profileId, config.connectTimeoutMs
        )
        if (printerHandle == 0L) {
            throw PrinterDriverException("pd_add_printer_tcp failed: $lastError")
        }
        return Printer(this, printerHandle)
    }

    /**
     * A printer reached over Bluetooth Classic SPP (docs/compatibility-brief.md §25),
     * using an `android.bluetooth.BluetoothSocket` this wrapper owns and drives.
     *
     * The device must already be paired -- pairing is an operator action through the
     * system's Bluetooth settings, and this SDK deliberately cannot perform it. The
     * calling app is responsible for holding `BLUETOOTH_CONNECT` and `BLUETOOTH_SCAN`
     * at runtime on API 31+; this library declares them but cannot request them, since
     * a runtime permission prompt needs an Activity.
     *
     * Throws [PrinterDriverException] on failure -- an unknown [BluetoothPrinterConfig.profileId]
     * or a vtable the core refused, both configuration mistakes. Note what is NOT a
     * failure here: the printer being out of range or switched off. Registration does not
     * connect; the core connects when it has something to send, and reports an
     * unreachable printer through the job's own [JobResult], which is where an operator
     * can act on it.
     *
     * NEVER RUN AGAINST HARDWARE. See the header comment of Bluetooth.kt for exactly what
     * has and has not been checked.
     *
     * @param adapter the system [BluetoothAdapter]; see the [Context] overload for the
     *   usual way to obtain one.
     */
    fun addPrinterBluetooth(adapter: BluetoothAdapter, config: BluetoothPrinterConfig): Printer {
        checkOpen()
        val transport = BluetoothSppTransport(adapter, config)
        val printerHandle = NativeBridge.addPrinterCustom(
            handle, transport, transport.description, config.profileId, config.widthDots
        )
        // Published even on failure: a reader thread the core may already have started
        // (see BluetoothSppTransport.attachTo) is waiting to be told, and 0L is how it
        // learns there is nothing to feed.
        transport.attachTo(handle, printerHandle)
        if (printerHandle == 0L) {
            throw PrinterDriverException("pd_add_printer_custom failed: $lastError")
        }
        bluetoothTransports[printerHandle] = transport
        return Printer(this, printerHandle)
    }

    /** [addPrinterBluetooth] with the adapter taken from the system service, which is
     *  the supported way to get one on the [minSdk][android.os.Build.VERSION_CODES.O]
     *  this library targets. Throws [PrinterDriverException] on a device with no
     *  Bluetooth hardware. */
    fun addPrinterBluetooth(context: Context, config: BluetoothPrinterConfig): Printer {
        val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
        val adapter = manager?.adapter
            ?: throw PrinterDriverException("this device has no Bluetooth adapter")
        return addPrinterBluetooth(adapter, config)
    }

    internal fun bluetoothTransportFor(printerHandle: Long): BluetoothSppTransport? =
        bluetoothTransports[printerHandle]

    /** Looks up any job this driver knows about, including ones reloaded from the
     *  journal after a restart (docs/api.md §2). `null` when [key] is unknown -- a
     *  routine, expected outcome, not an error. */
    /**
     * Paper to job (docs/api.md §14): resolves the four-character `V:` code printed on
     * a receipt.
     *
     * Accepts either of a job's identifiers, the print fence's or the cut fence's, and
     * answers most-recent-first -- the sequence wraps, and the receipt somebody is
     * holding is far more likely to be the recent one. Includes jobs reconstructed from
     * the journal, so a receipt printed before the last restart still resolves. `null`
     * when no job on this driver ever carried that token.
     */
    fun jobByToken(token: String): PrintJob? {
        checkOpen()
        val jobHandle = NativeBridge.jobByToken(handle, token)
        return if (jobHandle == 0L) null else PrintJob(this, jobHandle)
    }

    /**
     * The two characters every identifier this driver issues starts with: which driver
     * instance owns an echo, and therefore which instance printed a given receipt.
     * Persisted in the storage directory, so it survives a restart. A token that does
     * not start with this came from somewhere else -- the case
     * [DeviceEvent.FOREIGN_WRITER_DETECTED] reports.
     */
    val instanceNonce: String get() = NativeBridge.instanceNonce(handle)

    fun findJob(key: String): PrintJob? {
        checkOpen()
        val jobHandle = NativeBridge.findJob(handle, key)
        return if (jobHandle == 0L) null else PrintJob(this, jobHandle)
    }

    /**
     * Stops every printer worker, waits for in-flight jobs to reach a terminal state,
     * and frees every [Printer]/[PrintJob] handle this driver ever returned
     * (pd_destroy's contract, unchanged here). Idempotent -- a second call is a no-op.
     *
     * Cancels the closure-sugar scope first, then calls pd_destroy. Because
     * pd_job_await/pd_subscribe_job callbacks run on background threads that are not
     * cooperatively cancellable mid-native-call (see [PrintJob.result]'s KDoc), a
     * [Printer.send] callback that is already inside its native wait when [close] runs
     * may still fire once the shutdown-triggered terminal state comes through, even
     * though the scope has been asked to cancel. Callers that need a hard guarantee
     * that no callback fires after [close] returns should await outstanding results
     * first.
     *
     * Custom transports are torn down BEFORE pd_destroy, not after: pd_destroy frees
     * every pd_printer handle, and a Bluetooth reader thread still feeding one would be
     * writing through a dangling pointer. See [BluetoothSppTransport.shutdown] for what
     * that costs an in-flight job.
     */
    override fun close() {
        if (closedFlag.compareAndSet(false, true)) {
            closureScope.cancel()
            for (transport in bluetoothTransports.values) {
                transport.shutdown()
            }
            bluetoothTransports.clear()
            NativeBridge.driverDestroy(handle)
        }
    }

    internal fun checkOpen() {
        check(!closedFlag.get()) { "PrinterDriver is closed" }
    }

    internal fun launchClosureSugar(block: suspend CoroutineScope.() -> Unit) {
        checkOpen()
        closureScope.launch(block = block)
    }

    companion object {
        /** @throws PrinterDriverException if pd_create fails -- pd.h documents this as
         *  only possible when [PrinterDriverConfig.storageDirectory] could not be
         *  created; there is no driver handle yet at that point for pd_last_error to
         *  report against, so the message here is fixed rather than ABI-sourced. */
        fun create(config: PrinterDriverConfig = PrinterDriverConfig()): PrinterDriver {
            val nativeLogCallback = config.onLog?.let { callback -> NativeLogCallback { message -> callback(message) } }
            val handle = NativeBridge.driverCreate(config.storageDirectory, config.fsyncDisabled, nativeLogCallback)
            if (handle == 0L) {
                throw PrinterDriverException(
                    "pd_create returned null; the only documented cause is a storage " +
                        "directory that could not be created (see pd.h's pd_create doc)"
                )
            }
            return PrinterDriver(handle)
        }
    }
}

File: wrappers/dotnet/PrinterDriver/PrinterDriver.cs
using System.Runtime.InteropServices;

namespace PrinterDriver;

/// <summary>
/// The SDK entry point: owns the durable job journal, every printer's worker thread, and
/// every handle this wrapper ever returns.
/// </summary>
/// <remarks>
/// <para>
/// Dispose it once, at application shutdown. <see cref="Dispose"/> stops every printer
/// worker, waits for in-flight jobs to reach a terminal state, and only then releases the
/// callback roots — see <see cref="CallbackRoots"/> for why the order matters.
/// </para>
/// <para>
/// Handles are owned by the driver and are never freed individually, so the same
/// underlying job always maps to the same <see cref="PrintJob"/> instance. Idempotency-key
/// dedupe is therefore visible as reference equality, which is what
/// <c>PrintJobTests</c> asserts.
/// </para>
/// </remarks>
public sealed partial class PrinterDriver : IDisposable
{
    private readonly object _gate = new();
    private readonly CallbackRoots _roots = new();
    private readonly Dictionary<nint, Printer> _printers = [];
    private readonly Dictionary<nint, PrintJob> _jobs = [];
    private nint _handle;

    private PrinterDriver(nint handle) => _handle = handle;

    /// <summary>Opens a driver.</summary>
    /// <param name="config">Storage and durability settings; null means in-memory.</param>
    /// <returns>The driver. Dispose it at shutdown.</returns>
    /// <exception cref="PrinterDriverException">
    /// The storage directory could not be created. That is the only reason
    /// <c>pd_create</c> can fail, and it reports no detail — there is no handle to hang
    /// one on.
    /// </exception>
    public static PrinterDriver Create(PrinterDriverConfig? config = null)
    {
        config ??= new PrinterDriverConfig();
        var directory = config.StorageDirectory is { Length: > 0 } value
            ? Marshal.StringToCoTaskMemUTF8(value)
            : 0;
        try
        {
            var native = new PdConfig
            {
                StorageDirectory = directory,
                FsyncDisabled = config.FsyncDisabled ? 1 : 0,
                Log = 0,
                LogContext = 0,
            };
            var handle = NativeMethods.pd_create(in native);
            if (handle == 0)
            {
                throw new PrinterDriverException(
                    "the driver could not be created; the usual cause is a storage " +
                    $"directory that cannot be created ('{config.StorageDirectory}')");
            }
            return new PrinterDriver(handle);
        }
        finally
        {
            if (directory != 0)
            {
                Marshal.FreeCoTaskMem(directory);
            }
        }
    }

    /// <summary>
    /// The capability profile ids <see cref="TcpPrinterConfig.ProfileId"/> accepts, read
    /// from the ABI rather than hardcoded.
    /// </summary>
    public static IReadOnlyList<string> ProfileIds
    {
        get
        {
            var array = NativeMethods.pd_profile_ids();
            if (array == 0)
            {
                return [];
            }
            var ids = new List<string>();
            for (var i = 0; ; i++)
            {
                var entry = Marshal.ReadIntPtr(array, i * nint.Size);
                if (entry == 0)
                {
                    break;
                }
                ids.Add(NativeMethods.ReadUtf8(entry));
            }
            return ids;
        }
    }

    /// <summary>Attaches a printer reached over raw TCP (port 9100 by default).</summary>
    /// <param name="config">Where the printer is and what it is.</param>
    /// <returns>The printer, owned by this driver.</returns>
    /// <exception cref="PrinterDriverException">The ABI refused the configuration.</exception>
    public Printer AddPrinterTcp(TcpPrinterConfig config)
    {
        ArgumentNullException.ThrowIfNull(config);
        var handle = Handle;
        var id = Marshal.StringToCoTaskMemUTF8(config.PrinterId ?? string.Empty);
        var host = Marshal.StringToCoTaskMemUTF8(config.Host);
        var profile = Marshal.StringToCoTaskMemUTF8(config.ProfileId ?? string.Empty);
        try
        {
            var native = new PdTcpConfig
            {
                PrinterId = id,
                Host = host,
                Port = config.Port,
                WidthDots = config.WidthDots,
                ProfileId = profile,
                ConnectTimeoutMs = config.ConnectTimeoutMs,
            };
            var printer = NativeMethods.pd_add_printer_tcp(handle, in native);
            if (printer == 0)
            {
                throw new PrinterDriverException(LastError());
            }
            return Intern(printer);
        }
        finally
        {
            Marshal.FreeCoTaskMem(id);
            Marshal.FreeCoTaskMem(host);
            Marshal.FreeCoTaskMem(profile);
        }
    }

    /// <summary>
    /// Attaches a printer reached over a link the caller owns — Bluetooth Classic SPP, BLE,
    /// MFi, a vendor SDK channel, a USB bulk pipe, a test double.
    /// </summary>
    /// <remarks>
    /// docs/compatibility-brief.md §25. Read the thread contract on
    /// <see cref="IPrinterTransport"/> before implementing one: the core calls
    /// <see cref="IPrinterTransport.Connect"/>, <see cref="IPrinterTransport.Write"/> and
    /// <see cref="IPrinterTransport.Close"/> from its own worker thread for this printer,
    /// and received bytes come back in through
    /// <see cref="CustomTransportPrinter.FeedBytes(ReadOnlySpan{byte})"/> from a different
    /// one.
    /// </remarks>
    /// <param name="transport">The link. Kept alive by this driver until it is disposed.</param>
    /// <param name="config">Description, profile and width; null takes the defaults.</param>
    /// <returns>The printer, together with the two calls that feed the link's traffic in.</returns>
    /// <exception cref="PrinterDriverException">The ABI refused the registration.</exception>
    public CustomTransportPrinter AddPrinterCustom(IPrinterTransport transport,
                                                   CustomPrinterConfig? config = null)
    {
        ArgumentNullException.ThrowIfNull(transport);
        config ??= new CustomPrinterConfig();
        var handle = Handle;

        // Rooted before the function pointers are taken and never released until
        // pd_destroy: the stub Marshal.GetFunctionPointerForDelegate hands out lives
        // exactly as long as the delegate it came from, and the core keeps calling through
        // it for the whole life of the printer -- across reconnects, and from a thread no
        // managed reference reaches. This is the same defence as the subscription
        // callbacks, applied to the one place where a collected delegate would be called
        // while a receipt is half-written.
        var binding = _roots.Root(new TransportBinding(transport));
        var connect = new NativeMethods.TransportConnectCallback(
            TransportBinding.ConnectTrampoline);
        var write = new NativeMethods.TransportWriteCallback(TransportBinding.WriteTrampoline);
        var close = new NativeMethods.TransportCloseCallback(TransportBinding.CloseTrampoline);
        _roots.Root(connect);
        _roots.Root(write);
        _roots.Root(close);

        var description = Marshal.StringToCoTaskMemUTF8(config.Description ?? string.Empty);
        var profile = Marshal.StringToCoTaskMemUTF8(config.ProfileId ?? string.Empty);
        try
        {
            var vtable = new PdTransportVtable
            {
                Connect = Marshal.GetFunctionPointerForDelegate(connect),
                Write = Marshal.GetFunctionPointerForDelegate(write),
                Close = Marshal.GetFunctionPointerForDelegate(close),
                Description = description,
            };
            var printer = NativeMethods.pd_add_printer_custom(
                handle, in vtable, GCHandle.ToIntPtr(binding), profile, config.WidthDots);
            if (printer == 0)
            {
                throw new PrinterDriverException(LastError());
            }
            return new CustomTransportPrinter(this, Intern(printer));
        }
        finally
        {
            Marshal.FreeCoTaskMem(description);
            Marshal.FreeCoTaskMem(profile);
        }
    }

    /// <summary>
    /// Any job this driver knows about, including ones reloaded from the journal after a
    /// restart.
    /// </summary>
    /// <param name="key">The idempotency key.</param>
    /// <returns>The job, or null when the key is unknown.</returns>
    public PrintJob? FindJob(string key)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        var job = NativeMethods.pd_find_job(Handle, key);
        return job == 0 ? null : InternJob(job);
    }

    /// <summary>
    /// The path the native library was loaded from, for diagnosing a deployment that
    /// bound the wrong build.
    /// </summary>
    public static string? NativeLibraryPath => NativeLibraryResolver.ResolvedPath;

    internal nint Handle
    {
        get
        {
            var handle = Volatile.Read(ref _handle);
            ObjectDisposedException.ThrowIf(handle == 0, this);
            return handle;
        }
    }

    internal CallbackRoots Roots => _roots;

    internal string LastError()
    {
        var handle = Volatile.Read(ref _handle);
        if (handle == 0)
        {
            return "the driver has been disposed";
        }
        var message = NativeMethods.ReadUtf8(NativeMethods.pd_last_error(handle));
        return message.Length == 0 ? "the ABI reported no reason" : message;
    }

    internal Printer Intern(nint printer)
    {
        lock (_gate)
        {
            if (_printers.TryGetValue(printer, out var existing))
            {
                return existing;
            }
            var wrapper = new Printer(this, printer);
            _printers[printer] = wrapper;
            return wrapper;
        }
    }

    /// <summary>
    /// One <see cref="PrintJob"/> per <c>pd_job*</c>, forever. pd.h guarantees the same
    /// underlying job always maps to the same pointer, so interning here is what turns
    /// key dedupe into reference equality on the managed side.
    /// </summary>
    internal PrintJob InternJob(nint job)
    {
        lock (_gate)
        {
            if (_jobs.TryGetValue(job, out var existing))
            {
                return existing;
            }
            var wrapper = new PrintJob(this, job);
            _jobs[job] = wrapper;
            return wrapper;
        }
    }

    /// <summary>
    /// Stops every worker, waits for in-flight jobs to become terminal, then releases the
    /// callback roots.
    /// </summary>
    public void Dispose()
    {
        var handle = Interlocked.Exchange(ref _handle, 0);
        if (handle == 0)
        {
            return;
        }
        // Order matters and is not negotiable: pd_destroy blocks until every worker
        // thread has stopped, which is the moment after which no callback can run. Only
        // then is it safe to drop the delegates those threads were calling.
        NativeMethods.pd_destroy(handle);
        _roots.FreeAll();
        lock (_gate)
        {
            foreach (var job in _jobs.Values)
            {
                job.OnDriverDisposed();
            }
            _jobs.Clear();
            _printers.Clear();
        }
    }
}
